feature/s3/transfermanager/api_op_GetObject.go

```diff
@@ -0,0 +1,940 @@
+	}
+
+	var out *s3.GetObjectOutput
+	var n int64
+	var err error
+	for retry := 0; retry < d.options.PartBodyMaxRetries; retry++ {
+		out, n, err = d.tryDownloadChunk(ctx, params, &chunk, clientOptions...)
+		if err == nil {
+			break
+		}
+		// Check if the returned error is an errReadingBody.
+		// If err is errReadingBody this indicates that an error
+		// occurred while copying the http response body.
+		// If this occurs we unwrap the err to set the underlying error
+		// and attempt any remaining retries.
+		if bodyErr, ok := err.(*errReadingBody); ok {
+			err = bodyErr.Unwrap()
+		} else {
+			return nil, err
+		}
+
+		chunk.cur = 0
+
+		d.options.Logger.Logf(logging.Debug,
+			"object part body download interrupted %s, err, %v, retrying attempt %d",
+			aws.ToString(params.Key), err, retry)
+	}
+
+	d.incrWritten(n)
+
+	var output *GetObjectOutput
+	if out != nil {
+		output = &GetObjectOutput{}
+		output.mapFromGetObjectOutput(out, params.ChecksumMode)
+	}
+	return output, err
+}
+
+func (d *downloader) tryDownloadChunk(ctx context.Context, params *s3.GetObjectInput, chunk *dlchunk, clientOptions ...func(*s3.Options)) (*s3.GetObjectOutput, int64, error) {
+	out, err := d.options.S3.GetObject(ctx, params, clientOptions...)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	d.setTotalBytes(out) // Set total if not yet set.
+
+	var n int64
+	defer out.Body.Close()
+	if chunk.w != nil {
+		n, err = io.Copy(chunk, out.Body)
+		if err != nil {
+			return nil, 0, &errReadingBody{err: err}
+		}
+	} else {
+		part, err := ioutil.ReadAll(out.Body)
+		if err != nil {
+			return nil, 0, &errReadingBody{err: err}
+		}
+		n = int64(len(part))
+		d.buf = append(d.buf, part...)
+	}
+
+	return out, n, nil
+}
+
+func (d *downloader) incrWritten(n int64) {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	d.written += n
+}
+
+// getTotalBytes is a thread-safe getter for retrieving the total byte status.
+func (d *downloader) getTotalBytes() int64 {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	return d.totalBytes
+}
+
+// setTotalBytes is a thread-safe setter for setting the total byte status.
+// Will extract the object's total bytes from the Content-Range if the file
+// will be chunked, or Content-Length. Content-Length is used when the response
+// does not include a Content-Range. Meaning the object was not chunked. This
+// occurs when the full file fits within the PartSize directive.
+func (d *downloader) setTotalBytes(resp *s3.GetObjectOutput) {
+	d.m.Lock()
+	defer d.m.Unlock()
+
+	if d.totalBytes >= 0 {
+		return
+	}
+
+	if resp.ContentRange == nil {
+		// ContentRange is nil when the full file contents is provided, and
+		// is not chunked. Use ContentLength instead.
+		if aws.ToInt64(resp.ContentLength) > 0 {
+			d.totalBytes = aws.ToInt64(resp.ContentLength)
+			return
+		}
+	} else {
+		parts := strings.Split(*resp.ContentRange, "/")
+
+		total := int64(-1)
+		var err error
+		totalStr := parts[len(parts)-1]
+
+		if totalStr != "*" {
```

Comment:
Under which scenarios will this be `"*"`? Also, when this happens, `total` would be set to `-1`. Is that the desired behavior, or should we throw an error?

Reply:
Forgot to remove that check, s3 will always return content length